keanu-project/src/main/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertex.java

```diff
@@ -23,6 +25,27 @@
         return new CategoricalVertex<>(defensiveCopy(selectableValues));
     }
 
+    public static <T> CategoricalVertex<T> of(DirichletVertex vertex, List<T> categories) {
+        final int length = (int) TensorShape.getLength(vertex.getShape());
+        if (length != categories.size()) {
+            throw new IllegalArgumentException("Categories must have length of vertex's size");
+        }
+
+        final int categoriesCount = categories.size();
+        final IntStream categoriesIndices = IntStream.range(0, categoriesCount);
+        final Map<T, DoubleVertex> selectableValues =
+            categoriesIndices.boxed().collect(toMap(categories::get,
+                index -> new TakeVertex(vertex, 0, index)));
+        return new CategoricalVertex<>(selectableValues);
+    }
+
+    public static CategoricalVertex<Integer> of(DirichletVertex vertex) {
+        final int categoriesCount = (int) TensorShape.getLength(vertex.getShape());
```

Comment:
With this and below, we're limited by the size of an `int` (lists are indexed by `int`), so the cast has to happen at some point. My goal here was to make this as explicit as possible by casting early on.

Reply:
Some options:

- Defensively check for the length of a tensor being larger than an int (as this is very unlikely given a categorical)
- Allow for these longer tensors by changing the implementation to not go through the factory method that uses a `List`, possibly have a method that accepts a stream instead?
- Do nothing. It's unlikely enough to not matter

I think the first option makes the most sense.

Reply:
There is no cast if you do it the way I'm proposing. You're just iterating (or streaming) over some categories and putting them into a Collection.

It's true that the maximum size of a Collection is `Integer.MAX_VALUE` and something bad would happen if you tried to add more than that into a Collection, but we can reasonably assume that's never going to happen.

Reply:
Re: "Defensively check for the length of a tensor being larger than an int" - you're already doing that when you check `if (length != categories.size())`

Reply:
This check isn't happening in the other factory method though (with no list of categories). I'll add the check there to prevent silent weird behaviour and hopefully we'll be in a good place then.

Reply:
To be clear, the reason I don't want to return a `CategoricalVertex<Long>` is because it's exposing this issue to people consuming the API - they'll likely be using an int (as everything in this must be less than the size of an int), so they'll have to do these awkward casts. With the cast happening internally, we don't expose this. I'll try to make this clearer with comments.